simplified-books-database/src/main/java/org/nypl/simplified/books/book_database/DatabaseFormatHandleAudioBook.kt

```diff
@@ -293,23 +301,39 @@ internal class DatabaseFormatHandleAudioBook internal constructor(
     this.parameters.onUpdated.invoke(newFormat)
   }
 
-  override fun savePlayerPosition(position: PlayerPosition) {
-    val text =
-      JSONSerializerUtilities.serializeToString(PlayerPositions.serializeToObjectNode(position))
-
+  override fun setBookmarks(bookmarks: List<Bookmark.AudiobookBookmark>) {
     val newFormat = synchronized(this.dataLock) {
-      FileUtilities.fileWriteUTF8Atomically(this.filePosition, this.filePositionTmp, text)
-      this.formatRef = this.formatRef.copy(position = position)
+      FileUtilities.fileWriteUTF8Atomically(
+        this.fileBookmarks,
+        this.fileBookmarksTmp,
+        BookmarkJSON.serializeAudiobookBookmarksToString(this.parameters.objectMapper, bookmarks)
+      )
+      this.formatRef = this.formatRef.copy(bookmarks = bookmarks)
       this.formatRef
     }
 
     this.parameters.onUpdated.invoke(newFormat)
   }
 
-  override fun clearPlayerPosition() {
+  override fun setLastReadLocation(bookmark: Bookmark.AudiobookBookmark?) {
+
     val newFormat = synchronized(this.dataLock) {
-      FileUtilities.fileDelete(this.filePosition)
-      this.formatRef = this.formatRef.copy(position = null)
+      if (bookmark != null) {
+        Preconditions.checkArgument(
+          bookmark.kind == BookmarkKind.BookmarkLastReadLocation,
+          "Must use a last-read-location bookmark"
+        )
+
+        FileUtilities.fileWriteUTF8Atomically(
+          this.filePosition,
+          this.filePositionTmp,
+          BookmarkJSON.serializeAudiobookBookmarkToString(this.parameters.objectMapper, bookmark)
```

Comment:
Regarding the first question, the older versions of the app don't support server synced bookmarks, so there isn't a real case where a older version of the app would have to support the newer version saved positions because there's no way to downgrade the app. Regarding the second question, I'll test it and make the needed changes, if any

Reply:
What I meant was that the first part of your comment won't happen because it's not possible to set a local audiobook position in the newest version and then see if it's there in the oldest version because the user can't downgrade the app and even if he can, he'll have to uninstall the app and lose the saved position. However, there's indeed an issue with the second part of your question because the newest version of the app won't be able to handle the oldest version's saved position! I'll make the changes

Reply:
Oh, I get it now. That totally makes sense!